File: lib/faye/protocol/channel.rb
module Faye
  class Channel

    include Publisher
    attr_reader :name

    def initialize(name)
      super()
      @name = name
    end

    def <<(message)
      trigger(:message, message)
    end

    def unused?
      listener_count(:message).zero?
    end

    HANDSHAKE   = '/meta/handshake'
    CONNECT     = '/meta/connect'
    SUBSCRIBE   = '/meta/subscribe'
    UNSUBSCRIBE = '/meta/unsubscribe'
    DISCONNECT  = '/meta/disconnect'

    META        = 'meta'
    SERVICE     = 'service'

    class << self
      def expand(name)
        segments = parse(name)
        channels = ['/**', name]

        copy = segments.dup
        copy[copy.size - 1] = '*'
        channels << unparse(copy)

        1.upto(segments.size - 1) do |i|
          copy = segments[0...i]
          copy << '**'
          channels << unparse(copy)
        end

        channels
      end

      def valid?(name)
        Grammar::CHANNEL_NAME =~ name or
        Grammar::CHANNEL_PATTERN =~ name
      end

      def parse(name)
        return nil unless valid?(name)
        name.split('/')[1..-1]
      end

      def unparse(segments)
        '/' + segments.join('/')
      end

      def meta?(name)
        segments = parse(name)
        segments ? (segments.first == META) : nil
      end

      def service?(name)
        segments = parse(name)
        segments ? (segments.first == SERVICE) : nil
      end

      def subscribable?(name)
        return nil unless valid?(name)
        not meta?(name) and not service?(name)
      end
    end

    class Set
      def initialize
        @channels = {}
      end

      def keys
        @channels.keys
      end

      def remove(name)
        @channels.delete(name)
      end

      def has_subscription?(name)
        @channels.has_key?(name)
      end

      def subscribe(names, subscription)
        names.each do |name|
          channel = @channels[name] ||= Channel.new(name)
          channel.bind(:message, &subscription)
        end
      end

      def unsubscribe(name, subscription)
        channel = @channels[name]
        return false unless channel
        channel.unbind(:message, &subscription)
        if channel.unused?
          remove(name)
          true
        else
          false
        end
      end

      def distribute_message(message)
        channels = Channel.expand(message['channel'])
        channels.each do |name|
          channel = @channels[name]
          channel.trigger(:message, message) if channel
        end
      end
    end

  end
end

File: src/protocol/channel.js
'use strict';

var Class     = require('../util/class'),
    extend    = require('../util/extend'),
    Publisher = require('../mixins/publisher'),
    Grammar   = require('./grammar');

var Channel = Class({
  initialize: function(name) {
    this.id = this.name = name;
  },

  push: function(message) {
    this.trigger('message', message);
  },

  isUnused: function() {
    return this.countListeners('message') === 0;
  }
});

extend(Channel.prototype, Publisher);

extend(Channel, {
  HANDSHAKE:    '/meta/handshake',
  CONNECT:      '/meta/connect',
  SUBSCRIBE:    '/meta/subscribe',
  UNSUBSCRIBE:  '/meta/unsubscribe',
  DISCONNECT:   '/meta/disconnect',

  META:         'meta',
  SERVICE:      'service',

  expand: function(name) {
    var segments = this.parse(name),
        channels = ['/**', name];

    var copy = segments.slice();
    copy[copy.length - 1] = '*';
    channels.push(this.unparse(copy));

    for (var i = 1, n = segments.length; i < n; i++) {
      copy = segments.slice(0, i);
      copy.push('**');
      channels.push(this.unparse(copy));
    }

    return channels;
  },

  isValid: function(name) {
    return Grammar.CHANNEL_NAME.test(name) ||
           Grammar.CHANNEL_PATTERN.test(name);
  },

  parse: function(name) {
    if (!this.isValid(name)) return null;
    return name.split('/').slice(1);
  },

  unparse: function(segments) {
    return '/' + segments.join('/');
  },

  isMeta: function(name) {
    var segments = this.parse(name);
    return segments ? (segments[0] === this.META) : null;
  },

  isService: function(name) {
    var segments = this.parse(name);
    return segments ? (segments[0] === this.SERVICE) : null;
  },

  isSubscribable: function(name) {
    if (!this.isValid(name)) return null;
    return !this.isMeta(name) && !this.isService(name);
  },

  Set: Class({
    initialize: function() {
      this._channels = {};
    },

    getKeys: function() {
      var keys = [];
      for (var key in this._channels) keys.push(key);
      return keys;
    },

    remove: function(name) {
      delete this._channels[name];
    },

    hasSubscription: function(name) {
      return this._channels.hasOwnProperty(name);
    },

    subscribe: function(names, subscription) {
      var name;
      for (var i = 0, n = names.length; i < n; i++) {
        name = names[i];
        var channel = this._channels[name] = this._channels[name] || new Channel(name);
        channel.bind('message', subscription);
      }
    },

    unsubscribe: function(name, subscription) {
      var channel = this._channels[name];
      if (!channel) return false;
      channel.unbind('message', subscription);

      if (channel.isUnused()) {
        this.remove(name);
        return true;
      } else {
        return false;
      }
    },

    distributeMessage: function(message) {
      var channels = Channel.expand(message.channel);

      for (var i = 0, n = channels.length; i < n; i++) {
        var channel = this._channels[channels[i]];
        if (channel) channel.trigger('message', message);
      }
    }
  })
});

module.exports = Channel;
